File: src/graph/Dijkstra.java
package graph;

import java.util.*;
import java.util.stream.Stream;

public class Dijkstra {

    public static class Edge {
        final Integer to;
        final Integer weight;

        public Edge(Integer to, Integer weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    static class VertexPQ {
        final Integer weight;
        final Set<Integer> path;
        final Integer last;

        public VertexPQ(Integer weight, Set<Integer> path, Integer last) {
            this.weight = weight;
            this.path = path;
            this.last = last;
        }
    }

    public static Integer[] process(Integer start, Integer finish, Edge[][] graph) {
        // create heap with comparator
        PriorityQueue<VertexPQ> pq = new PriorityQueue<>((v1, v2) -> v1.weight - v2.weight);
        Set<Integer> objects = new HashSet<>();
        objects.add(start);
        pq.add(new VertexPQ(0, objects, start));
        // reduce with pq
        while (!pq.isEmpty()) {
            VertexPQ curr = pq.poll();
            Integer last = curr.last;
            Set<Integer> path = curr.path;
            if (finish.equals(last)) {
                // todo reorder from start to finish
                return path.toArray(new Integer[path.size()]);
            }
            // else
            Arrays.stream(graph[last])
                .filter(edge -> !path.contains(edge.to))
                    .map(edge ->
                            new VertexPQ(curr.weight + edge.weight,
                                    extendPath(curr, edge),
                                    edge.to))
                .forEach(pq::add);
        }
        return null;
    }

    private static Set<Integer> extendPath(VertexPQ curr, Edge edge) {
        Set<Integer> newPath = new HashSet<>();
        newPath.addAll(curr.path);
        newPath.add(edge.to);
        return newPath;
    }

}

File: src/Dijkstra.scala
import scala.collection.mutable

object Dijkstra {

  case class Edge (to: Int, weight: Int)

  case class VertexPQ(weight: Int, path: Set[Int], last: Int) extends Ordered[VertexPQ] {

    def compare(that: VertexPQ): Int = this.weight compare that.weight

  }

  def process(start: Int, finish: Int, graph: Array[Array[Edge]]): Option[Array[Int]] = {
    var pq = mutable.PriorityQueue[VertexPQ]()
    pq += VertexPQ(0, Set(start), start)
    while (pq.nonEmpty) {
      val curr: VertexPQ = pq.dequeue()
      val last: Int = curr.last
      val path: Set[Int] = curr.path
      if (finish == last) {
         return Some(path.toArray)
      }
      graph(last)
        .filter(edge => !path.contains(edge.to))
        .map(edge =>
            VertexPQ(curr.weight + edge.weight, curr.path + edge.to, edge.to))
        .foreach(pq += _)
    }
    None
  }

}
